validator/src/journal/block_manager.rs

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use std::collections::HashMap;
+
+use block::Block;
+use journal::NULL_BLOCK_IDENTIFIER;
+use journal::block_store::BlockStore;
+
+#[derive(Debug, PartialEq)]
+pub enum BlockManagerError {
+    MissingPredecessor(String),
+    MissingPredecessorInBranch,
+    MissingInput,
+    UnknownBlock,
+    UnknownBlockStore,
+}
+
+#[derive(Default)]
+struct Anchors {
+    anchors_by_blockstore_name: HashMap<String, Vec<String>>,
+
+    anchors_by_block_id: HashMap<String, Anchor>,
+}
+
+impl Anchors {
+    fn new() -> Self {
+        Anchors::default()
+    }
+
+    fn add_anchor(
+        &mut self,
+        block_id: &str,
+        blockstore_name: &str,
+        external_ref_count: i64,
+        internal_ref_count: i64,
+    ) {
+        let anchor = Anchor {
+            blockstore_name: blockstore_name.to_string(),
+            external_ref_count,
+            internal_ref_count,
+            block_id: block_id.to_string(),
```

Comment:
Per previous style discussions, prefer `into()` for string conversions like this.